### PR TITLE
block: fix FilterGUID

### PR DESCRIPTION
**Pull request by hugelgupf**
Hey @insomniacslk @tfg13 @pmazzini do y'all use this logic? Because it is broken as far as I can tell. Take a look at this small change emphasizing what this actually does.

Since the only usage of this doesn't do what the code seems to think it does -- can I get rid of this function and fix it so it returns the partition?

### CODE REVIEW

**Comment by mpictor** [comment]
I do have a need for filtering partitions by GUID in #1704. Looks like `storage.PartitionsByGUID` became `block.BlockDevices.FilterGUID` but didn't actually live up to its name?

**Comment by hugelgupf** [comment]
Yeah, never did. Let's implement a real one?

**Comment by insomniacslk** [comment]
I'll look at this shortly, thanks for letting me know :)

**Comment by insomniacslk** [comment]
@hugelgupf `FilterESP` is not used on our side, I implemented it with the intention to use it at some point but as far as I'm concerned it can be removed. Could you please also remove the comment about `FilterESP` in `cmds/boot/localboot/main.go` ?

**Comment by hugelgupf** [comment]
@insomniacslk my point was more that mountByGuid in localboot does not work as written. That means path mode doesn't work either. Or does it?

**Comment by insomniacslk** [comment]
On memory it should work, let me check on a laptop

**Comment by insomniacslk** [comment]
I'm not sure what's the mismatch between the function name/comment and its behaviour - it's looking for (the first) partition with a given GUID, mounts it, and returns the mount point as expressed in the function comment. What is the behaviour you expect?

**Comment by hugelgupf** [comment]
If you read the code of FilterGUID, and the test I wrote, you'll see it does not find a partition. It finds a DISK.

FilterGUID(ESP) = /dev/sda, NOT /dev/sda1

As written.

**Comment by hugelgupf** [comment]
What you describe is the behavior I expect, it's *not the behavior implemented*.

**Comment by tfg13** [comment]
It seems this was already wrong in the very first commit 2 years ago: https://github.com/u-root/u-root/commit/d74b17e8fedc4641f148bf7860f46809f6aba2ff, `FilterEFISystemPartitions` also returns the device instead of the partition, I suppose it went unnoticed because the function was essentially unused: https://github.com/u-root/u-root/commit/d74b17e8fedc4641f148bf7860f46809f6aba2ff#diff-b360ede5ae0435be5b7f7ce29d5ceb65R47

Not really sure how this went unnoticed for so long. I think we are not supplying a specific GUID, so it searches all partitions for grub configs, which then works fine and it never uses this function.

**Comment by hugelgupf** [comment]
Does that mean we can remove the path mode or do you want it fixed?

We should have a fixed FilterGUID function definitely, I'm just curious about what you want from localboot given this discovery.

**Comment by tfg13** [comment]
Should we just fix the function (to return the partition) and change the name back? From looking at `mountByGUID`, it does not look like it would work right now, or does it?

**Comment by insomniacslk** [comment]
> If you read the code of FilterGUID, and the test I wrote, you'll see it does not find a partition. It finds a DISK.
> 
> FilterGUID(ESP) = /dev/sda, NOT /dev/sda1
> 
> As written.

Sorry, I didn't catch what you meant - yes, it finds a disk, not a partition, and that's not the right behaviour. If someone else needs that, it should be fixed, otherwise we can remove it now and reintroduce it later

**Comment by tfg13** [comment]
I feel path mode might just work if we fix the function to actually return the partition

**Comment by hugelgupf** [comment]
It'll work, but it's obviously not used, so what's the gain in fixing it rather than deleting it? 

(I wanna fix it & delete code, unless you want to start using path mode some time)

**Comment by insomniacslk** [comment]
We are not going to use it soon, so this is OK to remove for me

**Comment by hugelgupf** [comment]
I fixed it, and I deleted localboot code. If you want to keep it, I'm fine with that, too -- just a proposal. PTAL

File: pkg/mount/block/blockdev.go [hunk 1]
```diff
@@ -387,32 +386,42 @@ func (b BlockDevices) FilterZeroSize() BlockDevices {
...
 
-// FilterESP returns a list of BlockDev objects whose underlying block device
-// is a valid EFI system partition.
-func (b BlockDevices) FilterESP() BlockDevices {
-	return b.FilterGUID(SystemPartitionGUID.String())
-}
-
-// FilterGUID returns a list of BlockDev objects whose underlying
-// block device has the given GPT partition GUID.
+// FilterGUID returns a list of partition BlockDev objects whose partition GUID
```
Comment (mpictor) [inline pkg/mount/block/blockdev.go:389]: I'd reword a bit:
```diff
-// FilterGUID returns a list of partition BlockDev objects whose partition GUID
+// FilterGUID returns a list of partition BlockDev objects whose partition type GUID
```

`part.Type` on line 402 is the partition's type guid, not its unique partition guid.

Comment (mpictor) [reply pkg/mount/block/blockdev.go:389]: I propose renaming this function, perhaps to `FilterPartType()` and creating a new one `FilterPartID()` which compares `part.Id` rather than `part.Type`.

Comment (hugelgupf) [reply pkg/mount/block/blockdev.go:389]: good point. changing it.

Comment (hugelgupf) [reply pkg/mount/block/blockdev.go]: done!

**Comment by hugelgupf** [comment]
> I fixed it, and I deleted localboot code. If you want to keep it, I'm fine with that, too -- just a proposal. PTAL

@insomniacslk @tfg13 i'd feel more comfortable merging this with one of your stamps of approval :)

**Comment by insomniacslk** [comment]
Right now this is safe to merge, I'm discussing internally when we will need path mode again (the "if we need it" being almost certainly "yes"), and will get back to this

**Comment by insomniacslk** [comment]
so - I'm reconsidering this a bit (and sorry for not verifying it earlier). Systemboot currently allows booting from grub config or from VPD variables (see [pkg/systembooter.GetBootEntries](https://github.com/u-root/u-root/blob/master/pkg/boot/systembooter/bootentry.go#L61) ). While removing these flags in localboot would not break that, it is still useful to have for manual troubleshooting (which we didn't need in a long time now).

Regarding the "it could have never worked", I'm baffled - I'm quite sure I've used this on my test devices, and I'm wondering if I messed up with the boot image (i.e. writing it to /dev/sda instead of /dev/sda1).

In conclusion, it is safe to merge this now, but I may need this feature in the future. If this is an easy fix here, I'd leave it, if not let's merge this, and I'll reintroduce it later when necessary. How does that sound?

**Comment by hugelgupf** [comment]
I'll revert the deletion. It should fix the localboot case.

**Comment by hugelgupf** [comment]
@mpictor PTAL -- now with FilterPartID :)

**Comment by insomniacslk** [comment]
Thank you Chris!
